File: src/main/java/com/bpbf/sirh_backend/services/TropPercuService.java
package com.bpbf.sirh_backend.services;

import com.bpbf.sirh_backend.dtos.TropPercuDto;
import com.bpbf.sirh_backend.entities.Employee;
import com.bpbf.sirh_backend.entities.SalaryElement;
import com.bpbf.sirh_backend.entities.TropPercu;
import com.bpbf.sirh_backend.repositories.EmployeeRepository;
import com.bpbf.sirh_backend.repositories.SalaryElementRepository;
import com.bpbf.sirh_backend.repositories.TropPercuRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class TropPercuService {

    private final TropPercuRepository tropPercuRepository;
    private final EmployeeRepository employeeRepository;
    private final SalaryElementRepository salaryElementRepository;

    @Transactional(readOnly = true)
    public List<TropPercuDto> getAll() {
        return tropPercuRepository.findAllByOrderByDateCreationDesc().stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TropPercuDto> getByEmployee(Long employeeId) {
        return tropPercuRepository.findByEmployeeId(employeeId).stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TropPercuDto> getByMoisApplication(String moisApplication) {
        return tropPercuRepository.findByMoisApplication(moisApplication).stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public TropPercuDto getById(Long id) {
        return tropPercuRepository.findById(id)
                .map(this::toDto)
                .orElseThrow(() -> new RuntimeException("Trop-perçu non trouvé avec l'id: " + id));
    }

    @Transactional
    public TropPercuDto create(TropPercuDto dto) {
        if (dto.getEmployeeId() == null) {
            throw new IllegalArgumentException("L'identifiant de l'employé est obligatoire");
        }
        if (dto.getAmount() == null || dto.getAmount().compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Le montant du trop-perçu doit être supérieur à zéro");
        }
        if (dto.getMoisApplication() == null || dto.getMoisApplication().isBlank()) {
            throw new IllegalArgumentException("Le mois de déduction (application) est obligatoire");
        }

        SalaryElement element = null;
        if (dto.getSalaryElementId() != null) {
            element = salaryElementRepository.findById(dto.getSalaryElementId()).orElse(null);
        }

        TropPercu entity = TropPercu.builder()
                .employeeId(dto.getEmployeeId())
                .salaryElement(element)
                .moisOrigine(dto.getMoisOrigine())
                .moisApplication(dto.getMoisApplication())
                .amount(dto.getAmount())
                .motif(dto.getMotif())
                .statut(dto.getStatut() != null && !dto.getStatut().isBlank() ? dto.getStatut() : "EN_ATTENTE")
                .build();

        TropPercu saved = tropPercuRepository.save(entity);
        log.info("Nouveau trop-perçu enregistré (id={}) pour l'employé id={} d'un montant de {} FCFA (Application: {})",
                saved.getId(), saved.getEmployeeId(), saved.getAmount(), saved.getMoisApplication());
        return toDto(saved);
    }

    @Transactional
    public TropPercuDto update(Long id, TropPercuDto dto) {
        TropPercu entity = tropPercuRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Trop-perçu non trouvé avec l'id: " + id));

        if (dto.getSalaryElementId() != null) {
            SalaryElement element = salaryElementRepository.findById(dto.getSalaryElementId()).orElse(null);
            entity.setSalaryElement(element);
        } else {
            entity.setSalaryElement(null);
        }

        entity.setEmployeeId(dto.getEmployeeId());
        entity.setMoisOrigine(dto.getMoisOrigine());
        entity.setMoisApplication(dto.getMoisApplication());
        entity.setAmount(dto.getAmount());
        entity.setMotif(dto.getMotif());
        if (dto.getStatut() != null && !dto.getStatut().isBlank()) {
            entity.setStatut(dto.getStatut());
        }

        TropPercu updated = tropPercuRepository.save(entity);
        log.info("Trop-perçu mis à jour (id={})", updated.getId());
        return toDto(updated);
    }

    @Transactional
    public void delete(Long id) {
        if (!tropPercuRepository.existsById(id)) {
            throw new RuntimeException("Trop-perçu introuvable pour suppression (id: " + id + ")");
        }
        tropPercuRepository.deleteById(id);
        log.info("Trop-perçu supprimé (id={})", id);
    }

    @Transactional
    public TropPercuDto changerStatut(Long id, String statut) {
        TropPercu entity = tropPercuRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Trop-perçu non trouvé avec l'id: " + id));
        entity.setStatut(statut);
        return toDto(tropPercuRepository.save(entity));
    }

    private TropPercuDto toDto(TropPercu entity) {
        TropPercuDto dto = TropPercuDto.builder()
                .id(entity.getId())
                .employeeId(entity.getEmployeeId())
                .moisOrigine(entity.getMoisOrigine())
                .moisApplication(entity.getMoisApplication())
                .amount(entity.getAmount())
                .motif(entity.getMotif())
                .statut(entity.getStatut())
                .dateCreation(entity.getDateCreation())
                .build();

        if (entity.getEmployeeId() != null) {
            employeeRepository.findById(entity.getEmployeeId()).ifPresent(emp -> {
                String fullName = ((emp.getNom() != null ? emp.getNom() : "") + " " + (emp.getPrenom() != null ? emp.getPrenom() : "")).trim(); dto.setEmployeeName(fullName);
                dto.setMatricule(emp.getMatricule());
            });
        }

        if (entity.getSalaryElement() != null) {
            dto.setSalaryElementId(entity.getSalaryElement().getId());
            dto.setSalaryElementName(entity.getSalaryElement().getName());
            dto.setSalaryElementCode(entity.getSalaryElement().getCode());
        }

        return dto;
    }
}

File: sigrh_frontend/src/app/features/paie/services/trop-percu.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

export interface TropPercuModel {
  id?: number;
  employeeId: number;
  employeeName?: string;
  matricule?: string;
  salaryElementId?: number;
  salaryElementName?: string;
  salaryElementCode?: string;
  moisOrigine?: string; // ex: "01/2026"
  moisApplication: string; // ex: "02/2026"
  amount: number;
  motif?: string;
  statut?: 'EN_ATTENTE' | 'APPLIQUE' | 'ANNULE' | string;
  dateCreation?: string;
}

@Injectable({
  providedIn: 'root'
})
export class TropPercuService {
  private readonly apiUrl = `${environment.apiUrl}/trop-percus`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<TropPercuModel[]> {
    return this.http.get<TropPercuModel[]>(this.apiUrl);
  }

  getByEmployee(employeeId: number): Observable<TropPercuModel[]> {
    return this.http.get<TropPercuModel[]>(`${this.apiUrl}/employee/${employeeId}`);
  }

  getByMois(mois: string): Observable<TropPercuModel[]> {
    return this.http.get<TropPercuModel[]>(`${this.apiUrl}/mois/${mois}`);
  }

  getById(id: number): Observable<TropPercuModel> {
    return this.http.get<TropPercuModel>(`${this.apiUrl}/${id}`);
  }

  create(item: Partial<TropPercuModel>): Observable<TropPercuModel> {
    return this.http.post<TropPercuModel>(this.apiUrl, item);
  }

  update(id: number, item: Partial<TropPercuModel>): Observable<TropPercuModel> {
    return this.http.put<TropPercuModel>(`${this.apiUrl}/${id}`, item);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  changerStatut(id: number, statut: string): Observable<TropPercuModel> {
    return this.http.patch<TropPercuModel>(`${this.apiUrl}/${id}/statut`, { statut });
  }
}